#996: Process recipes in bulk

andymikulski opened the pull request:
Fixes #996 - Long listing page load times

- Replaces firing multiple `dispatch`ed actions with one 'bulk' handler action
  - Specifically, instead of firing a bunch of `recipeReceived`s, we fire one action with type `RECIPE_PAGE_RECEIVE` and process all the recipes in one go in the reducer.
- This approach cuts the measured front end 'load time' by ~40-60% and greatly improves the UX around the two components
- This PR contains the changes for bulk processing the recipe listing and recipe histories.

mythmon commented on recipe-server/client/control_new/state/app/recipes/reducers.js:
```diff
@@ -38,23 +38,31 @@ function history(state = new Map(), action) {
-        .remove('approved_revision');
+      recipes.forEach(receivedRecipe => {
+        newState = newState.set(receivedRecipe.get('id'), formatRecipe(receivedRecipe));
+      });
```
As a drive comment, Immutable's [`withMutations` helper](https://facebook.github.io/immutable-js/docs/#/Map/withMutations) might help speed this up. It temporarily makes the data structure directly mutable, so you could write this as:

```js
const recipes = fromJS(action.recipes.results);
return state.withMutations(mutState => {
  recipe.forEach(receivedRecipe => {
    mutState.set(receivedRecipe.get('id'), formatRecipe(receivedRecipe));
  });
});
```

rehandalal commented:
holding off on merging in case you wanted to go with @mythmon's suggestion.

andymikulski commented:
@rehandalal I'll switch that out now, shall I merge once that change is put in?

rehandalal commented on recipe-server/client/control_new/state/app/recipes/reducers.js:
```diff
@@ -38,23 +38,30 @@ function history(state = new Map(), action) {
+      const recipes = fromJS(action.recipes.results);
 
-      return state.set(action.recipe.id, recipe);
+      return state.withMutations(mutState => {
```
I think this is way weirder syntax. It's harder to read to me.

I was under the impression the `withMutations` stuff was going to apply to the `formatRecipe` and `formatRevision` methods.

mythmon replied:
This is what I had in mind. What do you find harder to read about it?

Applying this only to the `formatRecipe` and `formatRevision` functions would have some of the performance  benefits, but wrapping the entire reducer transaction in `withMutation` (like done here) is better.

rehandalal commented:
I believe the TC failure is unrelated.

rehandalal commented:
needs to have latest master merged in (or be rebased).

andymikulski commented:
I rebased onto master, and also added a small commit that adds `withMutations` to the `format[Recipe|Revision]` functions. @rehandalal , mind double checking https://github.com/mozilla/normandy/pull/1042/commits/f3e7b3506bfed1dcafe170547c418c50c328feb4 ?

rehandalal commented:
TC failure is unrelated and happening on master.
